targets/TARGET_Silicon_Labs/TARGET_EFM32/trng/sl_trng.c

```diff
@@ -0,0 +1,373 @@
+{
+    uint32_t val32;
+    int i, ret=0;
+    uint32_t ctrl = device->CONTROL;
+
+    /* Setup control register */
+    device->CONTROL = TRNG_CONTROL_ENABLE | TRNG_CONTROL_TESTEN |
+                      TRNG_CONTROL_BYPNIST | TRNG_CONTROL_BYPAIS31;
+
+    /* Apply software reset */
+    sl_trng_soft_reset(device);
+
+    /* Write test vector to the key register. */
+    sl_trng_set_key(device,
+                    (const unsigned char*)test_vector_conditioning_key);
+
+    /* Write test vector to the TESTDATA register */
+    for (i=0; i<TEST_VECTOR_CONDITIONING_INPUT_SIZE; i++)
+    {
+      sl_trng_write_test_data(device,
+                              test_vector_conditioning_input[i]);
+    }
+
+    for (i=0; i<TEST_VECTOR_CONDITIONING_OUTPUT_SIZE; i++)
+    {
+        /* Wait for data to become available in the FIFO. */
+        while ( 0 == device->FIFOLEVEL );
+        /* Read output from the conditioning function */
+        val32 = device->FIFO;
+        /* Compare with expected test vector. */
+        if (val32 != test_vector_conditioning_output[i])
+        {
+            /*
+            mbedtls_printf("Conditioning test failed. "
+                           "Test output word %d 0x%lx. Expected 0x%lx\n",
+                           i, val32, test_vector_conditioning_output[i]);
+            */
+            ret = SL_TRNG_ERR_CONDITIONING_TEST_FAILED;
+        }
+    }
+
+    /* Restore initial value of control register */
+    device->CONTROL = ctrl;
+
+    return ret;
+}
+
+static int sl_trng_check_status( TRNG_TypeDef *device )
+{
+    uint32_t status = device->STATUS;
+
+    if ( (status & (TRNG_STATUS_PREIF
+                    | TRNG_STATUS_REPCOUNTIF
+                    | TRNG_STATUS_APT64IF
+                    | TRNG_STATUS_APT4096IF
+                    | TRNG_STATUS_ALMIF)) == 0 )
+    {
+        /* No errors */
+        return 0;
+    }
+
+    if ( status & TRNG_STATUS_PREIF )
+    {
+        /* On a preliminary noise alarm we clear the FIFO and clear
+         * the alarm. The preliminary noise alarm is not critical. */
+        status &= ~TRNG_STATUS_PREIF;
+        device->STATUS = status;
+        sl_trng_clear_fifo(device);
+        return SL_TRNG_ERR_PRELIMINARY_NOISE_ALARM;
+    }
+    else
+    {
+        /* Clear alarm conditions by doing a TRNG soft reset. */
+        sl_trng_soft_reset( device );
+        if ( status & TRNG_STATUS_REPCOUNTIF )
+        {
+            return SL_TRNG_ERR_REPETITION_COUNT_TEST_FAILED;
+        }
+        if ( status & TRNG_STATUS_APT64IF )
+        {
+            return SL_TRNG_ERR_ADAPTIVE_PROPORTION_TEST_64_FAILED;
+        }
+        if ( status & TRNG_STATUS_APT4096IF )
+        {
+            return SL_TRNG_ERR_ADAPTIVE_PROPORTION_TEST_4096_FAILED;
+        }
+        if ( status & TRNG_STATUS_ALMIF )
+        {
+            return SL_TRNG_ERR_NOISE_ALARM;
+        }
+    }
+
+    return 0;
+}
+
+int sl_trng_check_entropy( TRNG_TypeDef *device )
```

Comment:
see previous comment, but I think this function has less impact than previous test funciton